barretenberg/cpp/src/barretenberg/eccvm/msm_builder.hpp

```diff
@@ -43,138 +45,138 @@ class ECCVMMSMMBuilder {
-    };
-
-    struct alignas(64) AdditionTrace {
-        Element p1;
-        Element p2;
-        Element p3;
-        bool predicate;
-        bool is_double;
-    };
-
     /**
      * @brief Computes the row values for the Straus MSM columns of the ECCVM.
      *
      * For a detailed description of the Straus algorithm and its relation to the ECCVM, please see
      * https://hackmd.io/@aztec-network/rJ5xhuCsn
      *
-     * @param msms
-     * @param point_table_read_counts
-     * @param total_number_of_muls
-     * @return std::vector<MSMState>
+     * @param msms A vector of vectors of ScalarMuls.
+     * @param point_table_read_counts Table of read counts to be populated.
+     * @param total_number_of_muls A mul op in the OpQueue adds up to two muls, one for each nonzero z_i (i=1,2).
+     * @param num_msm_rows
+     * @return std::vector<MSMRow>
      */
-    static std::vector<MSMState> compute_msm_state(const std::vector<bb::eccvm::MSM<CycleGroup>>& msms,
-                                                   std::array<std::vector<size_t>, 2>& point_table_read_counts,
-                                                   const uint32_t total_number_of_muls,
-                                                   const size_t num_msm_rows)
+    static std::tuple<std::vector<MSMRow>, std::array<std::vector<size_t>, 2>> compute_rows(
+        const std::vector<MSM>& msms, const uint32_t total_number_of_muls, const size_t num_msm_rows)
     {
-        // N.B. the following comments refer to a "point lookup table" frequently.
-        // To perform a scalar multiplicaiton of a point [P] by a scalar x, we compute multiples of [P] and store in a
-        // table: specifically: -15[P], -13[P], ..., -3[P], -[P], [P], 3[P], ..., 15[P] when we define our point lookup
-        // table, we have 2 write columns and 4 read columns when we perform a read on a given row, we need to increment
-        // the read count on the respective write column by 1 we can define the following struture: 1st write column =
-        // positive 2nd write column = negative the row number is a function of pc and slice value row = pc_delta *
-        // rows_per_point_table + some function of the slice value pc_delta = total_number_of_muls - pc
-        // std::vector<std::array<size_t, > point_table_read_counts;
-        const size_t table_rows = static_cast<size_t>(total_number_of_muls) * 8;
-        point_table_read_counts[0].reserve(table_rows);
-        point_table_read_counts[1].reserve(table_rows);
-        for (size_t i = 0; i < table_rows; ++i) {
+        // To perform a scalar multiplication of a point P by a scalar x, we precompute a table of points
+        //                           -15P, -13P, ..., -3P, -P, P, 3P, ..., 15P
+        // When we perform a scalar multiplication, x into four-bit scalars and conver them to wNAF form, then look
+        // these precomputed values up with. We record read counts in a table with the following structure:
```

Comment:
This sentence is a bit hard to read, type "conver" and I think you forgot decompose before x